src/image_streamer.cpp

```diff
@@ -24,13 +24,60 @@ ImageTransportImageStreamer::ImageTransportImageStreamer(const async_web_server_
   output_height_ = request.get_query_param_value_or_default<int>("height", -1);
   invert_ = request.has_query_param("invert");
   default_transport_ = request.get_query_param_value_or_default("default_transport", "raw");
-  qos_profile_name_ = request.get_query_param_value_or_default("qos_profile", "default");
+  qos_profile_name_ = request.get_query_param_value_or_default("qos_profile", "auto");
 }
 
 ImageTransportImageStreamer::~ImageTransportImageStreamer()
 {
 }
 
+std::optional<rmw_qos_profile_t> ImageTransportImageStreamer::detect_publisher_qos(const std::string &topic_name) 
+{
+  RCLCPP_INFO(nh_->get_logger(), "Attempting to auto-detect QoS for topic: %s", topic_name.c_str());
+
+  auto topic_endpoint_info_array = nh_->get_publishers_info_by_topic(topic_name);
+  if (topic_endpoint_info_array.empty()) {
+    RCLCPP_WARN(nh_->get_logger(), "No publishers found for topic: %s", topic_name.c_str());
+    return std::nullopt;
+  }
+
+  // Use the first publisher's QoS as reference.
+  auto endpoint_info = topic_endpoint_info_array.front();
+  auto qos_profile = endpoint_info.qos_profile();
+
+  // Log the detected QoS settings.
+  std::string reliability =
+      (qos_profile.reliability() == rclcpp::ReliabilityPolicy::Reliable) ? "RELIABLE" : "BEST_EFFORT";
+  std::string durability =
+      (qos_profile.durability() == rclcpp::DurabilityPolicy::TransientLocal) ? "TRANSIENT_LOCAL" : "VOLATILE";
+
+  RCLCPP_INFO(nh_->get_logger(), "Detected QoS - Reliability: %s, Durability: %s, History depth: %zu",
+              reliability.c_str(), durability.c_str(), qos_profile.depth());
+
+  // Convert rclcpp QoS to rmw QoS profile.
+  rmw_qos_profile_t rmw_qos = rmw_qos_profile_default;
+
+  // Set reliability
+  if (qos_profile.reliability() == rclcpp::ReliabilityPolicy::Reliable) {
+    rmw_qos.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+  } else {
+    rmw_qos.reliability = RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT;
+  }
+
+  // Set durability
+  if (qos_profile.durability() == rclcpp::DurabilityPolicy::TransientLocal) {
+    rmw_qos.durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
+  } else {
+    rmw_qos.durability = RMW_QOS_POLICY_DURABILITY_VOLATILE;
+  }
+
+  // Set history policy and depth
+  rmw_qos.history = RMW_QOS_POLICY_HISTORY_KEEP_LAST;
+  rmw_qos.depth = qos_profile.depth();
```

Comment:
Alternatively, you could organize it like this
```suggestion
  // Set defaults
  rmw_qos.reliability = RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT;
  rmw_qos.durability = RMW_QOS_POLICY_DURABILITY_VOLATILE;
  rmw_qos.history = RMW_QOS_POLICY_HISTORY_KEEP_LAST;
  rmw_qos.depth = qos_profile.depth();

  // Set reliability
  if (qos_profile.reliability() == rclcpp::ReliabilityPolicy::Reliable) {
    rmw_qos.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
  }
  // Set durability
  if (qos_profile.durability() == rclcpp::DurabilityPolicy::TransientLocal) {
    rmw_qos.durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
  }
```
But I don't have a strong argument here for one or the other, I just tend to avoid `else` statements.

Reply:
I like it.

Reply:
I think you pushed an old version of this function?  I remember seeing like 30  minutes ago this being changed and now it's back to the original